#+TITLE: Vidjil -- Browser Manual
#+AUTHOR: The Vidjil team (Mathieu, Mikaël and Marc)

This is the help of the Vidjil browser : [[http://rbx.vidjil.org/browser/]].
Further help can always be asked to [[mailto:contact@vidjil.org][contact@vidjil.org]]. We can also arrange phone or Skype meeting.

The Vidjil team (Mathieu, Mikaël and Marc)

* Requirements

** Browser

The Vidjil browser runs in any modern browser. It has been successfully tested on the following platforms
 - Firefox version >= 32
 - Chrome version >= 38
 - IE version >= 10.0 (Vidjil will not run on IE 9.0 or below)
 - Opera version >= XX
 - Safari version >= 6.0

** The .vidjil files

The vidjil browser displays =.vidjil= files that summarize the V(D)J
rearrangements and the sequences found in a run. 

The easiest way to get these files is to [[http://rbx.vidjil.org/browser][request an account]] on the Vidjil server.
You will then be able to upload,
manage, process your runs (=.fasta=, =.fastq=, =.gz= or =.clntab= files) directly on the browser 
(see below 'patient database'), and the server behind the patient
database computes these .vidjil files.
Otherwise, such .vidjil files can be obtained:
 - from the command-line version of Vidjil (starting from
   =.fasta= or =.fastq= files, see [[http://git.vidjil.org/blob/master/doc/algo.org][algo.org]]).
   To gather several .vidjil files, you have to use the [[http://git.vidjil.org/blob/master/server/fuse.py][fuse.py]] script
 - or by post-processing of other V(D)J analysis pipelines (contact us
   if you are interested)


* First aid

- Open data by:
    - either with “patients”/“open patient”  (if you have access to the database)
    - or with “file”/“import/export”, manually selecting a .vidjil file
  There is always a “LIL-L2” dataset for demonstration purposes.

- You can change the number of displayed clones by moving the slider “number of clones” (menu “filter”).
  The maximal number of clones that can be displayed depends on the processing step before.
  See below ("Can I see all the clones ?").

- Due to sequencing errors, there may be several clones corresponding to a real clone. 
   - You can select several clones (for example those sharing a same V and a same J), 
   - inspect the sequences in the lower panel (possibly using the “align” function),
   - and click on “merge” if you think that the clones should be merged. 
     Once several clones are merged, you can still visualize them by clicking on “+” in the list of clones.


* The elements of the Vidjil browser

** The info panel (upper left panel)
   - analysis :: name of the configuration file used for displaying the data
   - system :: system used for analyzing the data. In case of multi-system
               data, you can select what system should be displayed.
   - point :: name of the current point (you can change the selected point by clicking on
              another point in the graph). The name can be edited (“...” button).
   - date :: when the run was performed (manually curated, with “...” button)
   - segmented :: number of reads where Vidjil found a CDR3, for that point.
                  See [[Number of segmented reads]] below.
   - total :: total number of reads for that point

** The list of clones (left panel)

- You can assign other tags (and thus colors) to clones using the “★” button.
  The “filter” menu allows to further filter clones by tags.
- Under the “★” button it is possible to normalize clone concentrations
  according to this clone. You must specify the expected concentration in the
  “expected size” field (e.g. 0.01 for 1%). See [[Control with standard/spike]] below.

- The “i” button displays additional information on each clone.

- The list can be sorted on V genes, J genes or concentrations. At the top of
  the list you need to click respectively on “V sort”, “J sort” or “sort”.
  The “+” and “-” allow respectively to un-merge or re-merge all clones that have
  already been merged.

** The graph

- The gray areas at the bottom of the graph show, for each point, the resolution (1 read / 5 reads).

- You can reorder the points by dragging them, and hide some points by dragging them on the “+” mark at the right of the points.
  If you want to recover some hidden points, you need to drag them from the “+” mark to the graph.

- If your dataset contains sampling dates (for example in a MRD setup), you can switch between point keys and dates in “settings > point key”

- The vertical gray area shows the current point, you can change that by clicking on another point.


** The scatterplot view

- The axes of the plot (by default “V gene” / “J gene”) can be changed.

- Some presets are available in the “analysis” menu.
  
  To segregate a set of clones sharing a same V and J, it is often useful
  to display the clones according to their “N length” (that is N1-D-N2 in the case of VDJ rearrangements)

** The aligner (bottom panel)
   - When several clones are selected (you can select clones by clicking on
     them either in the list, the graph or the scatterplot, or by drawing a
     rectangle around clones to be selected in the scatterplot view), you can
     view their sequences in the aligner.
   - Sequences can be aligned together to see how they differ or how similar
     they are (“align” button). After aligning them a shaded background identifies
     substitutions and a dash identifies indels.
   - You can remove sequences from the aligner by clicking on their name (and
     therefore, you unselect them).
   - You can visualize results by IMGT/V-QUEST and IgBlast on the selected sequences, in another window, by clicking on the corresponding buttons.
   - You can unselect all sequences by clicking on the background of the scatterplot.


** The patient database and the server

If a server with a patient database is configured with your
installation of Vidjil (as on http://rbx.vidjil.org/browser), the
'patient' menu gives you access to the server.

With authentication, you can add patients, then add either
=.fasta=/=.fastq=/=.gz= files or =.clntab= files, then process your
runs and save the results of your analysis.

*** Patients
      
Once you are authentified, this page show the patient list. Here you
can see your patients and patients whose permission has been given to you.

New patients can be added ('add patient'), edited ('e') or deleted ('X').
By default, you are the only one who can see and update this new patient.
If you have an admin access, you can grant access to other users ('p').

*** Samples

Clicking on a patient give acccess the "samples" page. Each sample is
a =.fasta=/=.fastq=/=.gz=/=.clntab= file that will be processed by one or several
pipelines.
You can see which samples have been processed with the selected
config, and download the sequence files if they are available ("dl").

Depending on your granted accesses, you can 
add a new sample to the list ("add file"), 
schedule a processing for a sequence file (select a config and "run"),
or delete a sample ("X").

The processing can take a few seconds to a few hours, depending on the
software lauched, its options and the size of the sample.
Once the processing is finished, click on the button "see result" and
the browser will load the data of the processed files. The first click
on this button can take a few seconds.



* Can I see all the clones ?


The interest of NGS/Rep-Seq studies is to provide a deep view of any
V(D)J Repertoire. The underlying analysis software (such as Vidjil)
try to analyze as much reads as possible (see below 'Number of segmented reads').
One often wants to "see all clones", but a complete list is difficult
to see in itself. In a typical dataset with about 10^6 reads, even in
the presence of a dominant clone, there can be 10^4 or 10^5 different
clones detected.

** The "top" slider in the "filter" menu

The "top 10" clones are the clones that are in the first 10 ones
in *at least one* sample. As soon as one clone is in this "top 10"
list, it is displayed for every sample, even if its concentration is
very low in other samples.
Most of the time, a "top 10" is enough. The hidden clones are thus the
one that never reach the 10 first clones. With a default installation,
the slider can be set to display clones until the "top 100" on the grid 
(and, on the graph, until "top 20").

However, in some cames, one may want to track some clones that are
never in the "top 100", as for example:
  - a standard/spike with low concentration
  - a clone in a MRD following of a patient without the diagnostic point

(Upcoming feature). If clone is "tagged" in the =.vidjil= or
in the =.analysis= file, it will always be shown even if it does not
meet the "top" filter.

** The "other" clone

This virtual clone in the clone list groups all clones that are hidden
(because of the "top" or because of hiding some tags). The sum of
ratios in the list of clones is always 100%: thus the "other" clone
changes when one use the "filter" menu.

Note that the ratios include the "other" clone: if a clone principal
is reported to have 10.54%, this 10.54% ratio relates to the number of
analyzed reads, including the hidden clones.



* How can I assess the quality of the data and the analysis ?

To make sure that the PCR, the sequencing and the Vidjil analysis went well, several elements can be controlled.

** Number of segmented reads
A first control is to check the number of “segmented reads” in the info panel. For each point, this shows the number of reads where Vidjil found a CDR3. 
     
Ratios above 90% usually mean very good results. Smaller ratios, especially under 60%, often mean that something went wrong.
There can be several causes leading to bad ratios: 

*** analysis or biological causes

   - a system (for example TRG) was analyzed and the data actually contains other systems.
      (solution: ask that we relaunch Vidjil with other systems)

   - there are incomplete/exceptional rearrangements 
     (Vidjil can process some of them)

   - there are too many hypersomatic mutations
     (usually Vidjil can process mutations until 10% mutation rate... above that threshold, some sequences are lost)

*** PCR or sequencing causes

   - the read length is too short, the reads do not span the junction zone 
      (Vidjil detects a “window” including the CDR3. By default this window is 40–60bp long, so the read needs be that long)

   - In particular, for paired-end sequencing, one of the ends can lead to reads not fully containing the CDR3 region
      (solution: ignore this end, or extend the read length)

   - There were too many PCR or sequencing errors
      (this can be asserted by inspecting the related clones, checking if there is a large dispersion around the main clone)

** Control with standard/spike

   - If your sample included a standard/spike control, you should first
     identify the main standard sequence (if that is not already done) and
     specify its expected concentration (by clicking on the “★” button).
     Then the data is normalized according to that sequence.
   - You can (de)activate normalization in the settings menu.

** Steadiness verification
   - When assessing different PCR primers, PCR enzymes, PCR cycles, one may want to see how regular the concentrations are among the points.
   - When following a patient one may want to identify any clone that is emerging.
   - To do so, you may want to change the color system, in the “color” menu
     select “by abundance at selected timepoint”.  The color ranges from red
     (high concentration) to purple (low concentration) and allows to easily
     spot on the graph any large change in concentration.



* Browser API

The browser can be opened on a data file specified from a =data= attribute, 
and optionally on an analysis file specified from a =analysis= attribute,
as in the following URLs on our test server:

- http://rbx.vidjil.org/browser/?data=test.vidjil
- http://rbx.vidjil.org/browser/?data=test.vidjil&analysis=test.analysis
- http://rbx.vidjil.org/browser/?data=http://rbx.vidjil.org/browser/test.vidjil

Both GET and POST requests are accepted.
Note that the =browser/index.html= file and the =.vidjil/.analysis= files should be hosted on the same server.
Otherwise, the server hosting the =.vidjil/.analysis= files must accept cross-domain queries.


* Reference

If you use Vidjil for your research, please cite the following reference:

Mathieu Giraud, Mikaël Salson, et al.,
“Fast multiclonal clusterization of V(D)J recombinations from high-throughput sequencing”,
BMC Genomics 2014, 15:409 
http://dx.doi.org/10.1186/1471-2164-15-409


